feat(minimessage): named arguments

Strokkur424 opened the pull request:
closes #1296.

Strokkur424 committed:
feat: start work on named arguments in tags

Strokkur424 committed:
feat: modify token parser to parse named arguments

Strokkur424 committed:
chore: introduce new TokenType to uniquely distinguish value-less toggle arguments

Strokkur424 committed:
feat: (WIP) abstracting away TagProvider into QueuedTagProvider and NamedTagProvider and start adjusting MiniMessageParser to distinguish tags with named arguments

Strokkur424 committed:
fix: (WIP) parser should now theoretically be able to distinguish named arguments

Strokkur424 committed:
feat: flesh out parsing logic further and fix a bunch of issues

Strokkur424 committed:
feat: add test for basic named argument parsing

Strokkur424 committed:
feat: start adding more tests

Strokkur424 committed:
fix: <red > tag with space being recognized as valid tag

Strokkur424 committed:
feat: add a bunch more tests

Strokkur424 committed:
chore: fix all compile time issues

Strokkur424 committed:
feat: add test

Strokkur424 committed:
chore: cleanup diff and rename to sequential

Strokkur424 committed:
chore: add a bunch more tests

Strokkur424 committed:
feat: add inverted flag arguments

Strokkur424 committed:
feat: split the claiming resolvers into named and sequenced resolvers

Strokkur424 committed:
feat: add missing context newException method and try to parse tag with no arguments as named if no sequential resolver exists

Strokkur424 committed:
feat: add isFlagPresent

Strokkur424 committed:
fix: invalid tag in test

Strokkur424 commented:
I've rebased the PR to the `unstable/4` branch so that I can do some work for the upcoming head tag with the changes introduced here already.

Strokkur424 committed:
feat: add named argument support to token emitter

Strokkur424 committed:
feat: add flag support to token emitter

Strokkur424 committed:
chore: remove unused import

kezz commented on text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/resolver/TagResolver.java:
```diff
@@ -223,6 +273,18 @@ public boolean has(final @NotNull String name) {
+   * @throws ParsingException if the provided arguments are invalid
+   * @since 4.25.0
+   */
+  @Nullable Tag resolveNamed(@TagPattern final @NotNull String name, final @NotNull NamedArgumentMap arguments, final @NotNull Context ctx) throws ParsingException;
```
Needs to have a default null impl otherwise this is a breaking change

Strokkur424 committed:
chore: default implement TagResolver methods with null

Strokkur424 committed:
chore: remove instanceof check in SequentialTagResolver in order to not break backward compatibility

Strokkur424 committed:
chore: add missing (at)since annotations

jpenilla commented on text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/resolver/TagResolver.java:
```diff
@@ -142,6 +142,56 @@ public boolean has(final @NotNull String name) {
+   * Create a tag resolver that only responds to certain tag names, and whose value does not depend on that name.
+   *
+   * <p>
+   * This method creates a special resolver which listens to tags with named arguments instead of sequences ones.
```
```suggestion
   * This method creates a special resolver which listens to tags with named arguments instead of sequential ones.
```
there looks to be a couple other spots with the same issue

Strokkur424 replied:
I think I fixed them all

Strokkur424 committed:
chore: spelling mistakes

Strokkur424 commented:
Rebased to main/4 after the unstable/4 merge

zml2008 commented:
So a thought on this, why are we creating two different colours of MM tags? I feel like the design would be cleaner if we only had one type of tag, which accepted both named and sequential arguments, same as say with a CLI. For consistency's sake I think it's fair to require all the named args to be at the beginning or end of the tag, but either way there's a bunch of api noise and opportunity for breakages here due to adding named args in as a separate parallel concept.

I'm seeing something like:
```xml
<gradient phase=0.4 speed=2 red:green:blue>hi strokkur</gradient>
```
or:
```xml
<gradient:red:green:blue phase=0.4 speed=2>hi strokkur</gradient>
```

What do you think? Users would just pop named args from the `ArgumentQueue` same as other items

Strokkur424 committed:
Merge branch 'main/5' into feat/named-arguments

Strokkur424 committed:
chore: fix some main/5 rebase issues

Strokkur424 committed:
Merge branch 'main/5' into feat/named-arguments

Strokkur424 committed:
refactor: combine named and sequential tag arguments

KNOWN ISSUES/NOTES:
- *Yes*, I did indeed completely rewrite TokenParser.parseSecondPass. A few notes related to that:
  - This was needed because the previous system did not really play well with with state switches between named args and sequential ones and was honestly really hard to work with in general. My hopes are that with this rewrite it will be much more maintainable and readable.
  - Tags, which look like this <name > currently parse the same as <name>. I am awaiting a response regarding whether we want to make this the new expected behavior or if I need to do some special case handling for it.
  - Urls in values currently do not work. That's a quick fix though.

Strokkur424 committed:
fix: remaining tests

Strokkur424 committed:
refactor: minor second pass token parser changes and added banchmark method for tokenization

Strokkur424 committed:
chore: spotless apply

Strokkur424 committed:
chore: fix checkstyle violations and do a minor diff cleanup

Strokkur424 committed:
feat: add tests for combined arguments and for sequenced args with a space before

Strokkur424 committed:
chore: final cleanup

Strokkur424 commented:
Alright, I've done everything I've wanted to do for this PR now. Arguments have been implemented, as per @zml2008's suggestion, in a combined way. They new syntax looks like this: 

```xml
<tagname
    someflag
    somename='a quoted value with spaces'
    anotherflag :queued arguments behind:here>Hey abby</tagname>
```

This means named arguments are simply placed between the tag name and the first colon.

Strokkur424 committed:
Merge branch 'main/5' into feat/named-arguments

Strokkur424 committed:
Merge branch 'main/5' into feat/named-arguments
